### ISSUE TITLE
[Feature idea] linux support

### ISSUE THREAD

**Issue by OmgImAlexis** [berstend/node-safe#4]
Would love to see linux support.

**Comment by berstend** [comment]
Definitely! I mentioned the findings of my initial Linux research [in the readme](https://github.com/berstend/node-safe#other-platforms).

We basically need to find a suitable sandbox implementation that we want to instrument with node-safe on Linux. In the readme I listed our minimum sandbox requirements further down as well. 

It's a shame AppArmor dropped their `--file` option support a few years back, would've been perfect for our use-case 😄 

If anyone has an idea which Linux sandbox implementation would be the most suitable for us or does some further research please report your findings here :)

**Comment by alphaleadership** [comment]
and windows support

**Comment by alip** [comment]
Hello, I'd love to see node-safe on Linux too and as the main author of [sydbox](https://sydbox.exherbolinux.org/), I want to say sydbox mostly fits your purposes. Let me give a short review and pointers for further reading. Feel free to ping me for any questions:

> We're looking for a fast (no lengthy boot), non-root implementation we can control through the command line or environment. Minimum control we need is file system access (read/write separately), networking (at least outbound connections) and process forking. Ideally the filtering is file path based and supports extended (bash 4 like) globbing or regex.

1. Syd startup is relatively fast. E.g: On Exherbo, our package manager Paludis executes Syd once per each exheres (package definition file). That means syd is executed hundreds of times in a row. You can further improve startup time with `SYD_QUICK_BOOT=1` or `syd -q` which makes startup faster in return for sacrificing one layer of defense against some container breaks (ie memfd self-reexec, the next layer of defense is [procfs hardening](http://man.exherbolinux.org/syd.7.html#Hardened_procfs)).
2. Syd does not need root or any elevated privileges, and we will _never_ add a feature to Syd that requires elevated privileges. That's a design decision we made at step 0. Our goal is make security easy to use and accessible.
3. For filesystem access, Syd offers [Read](https://man.exherbolinux.org/syd.7.html#Read_Sandboxing), [Write](https://man.exherbolinux.org/syd.7.html#Write_Sandboxing), [Stat](https://man.exherbolinux.org/syd.7.html#Stat_Sandboxing), [Exec](http://man.exherbolinux.org/syd.7.html#Exec_Sandboxing) and [Ioctl](https://man.exherbolinux.org/syd.7.html#Ioctl_Sandboxing) sandboxing. All these sandboxing types accepts rsync-like globs as arguments and environment variables are expanded. More than one rule may be added and the last matching rule wins, e.g. `allow/read,stat+${HOME}/Desktop/**/*.png` or `allow/exec+/usr/**/bin/[bd]ash`. In addition there's [Lock](https://man.exherbolinux.org/syd.7.html#Lock_Sandboxing) which uses Landlock. The paths for lock sandboxing are checked at kernel level, hence they are prefixes, not globs. We have many more sandboxing types (such as TPE, SegvGuard and Binary Verification!), see the [_syd_(7)](https://man.exherbolinux.org/syd.7.html) manual page for more information.
4. For networking, Syd offers [Network](https://man.exherbolinux.org/syd.7.html#Network_Sandboxing) sandboxing where you can specify address patterns to bind or connect to using CIDR notation. We also have a convenience option `trace/allow_safe_bind:1` when Syd auto-allowlists successful bind addresses for subsequent connects.
5. To limit process forking under Syd, you have two main options Exec and [PID](https://man.exherbolinux.org/syd.7.html#PID_sandboxing) sandboxing, where you either limit what binaries can be executed or how many processes (threads actually) are allowed to co-exist. PID sandboxing can be thought as a simple alternative to PID Cgroups and is typically used with PID namespaces. e.g. if you do `syd -msandbox/pid:on -mpid/max:1 -munshare/user,pid:1 node...`, you effectively limit node to a single thread.
6. As I told before, rsync-like globs are supported. That should be adequate for your usecase.
7. Syd allows dynamic configuration of the sandbox from within the sandbox. The [API](https://man.exherbolinux.org/syd.2.html) is very simple and based on _stat_(2) calls. There're [C](https://libsyd.exherbolinux.org/), [Go](https://gosyd.exherbolinux.org/), [Perl](https://plsyd.exherbolinux.org/), [Python](https://pysyd.exherbolinux.org/), [Ruby](https://rbsyd.exherbolinux.org/), [Emacs Lisp](https://sydel.exherbolinux.org/) implementations of the Sydbox API. The C and Emacs Lisp implementations directly use the magic _stat_(2) calls. Other language bindings use the C library via FFI. Writing a pure Javascript version of the API should be trivial and give node-safe a lot of power in configuring the sandbox.
8. Last but not least, Sydbox has a learning mode that you can use with [Pandora](https://crates.io/crates/pandora_box)!

Here are a few more links if you want to learn further:
- [AsciiCast: PID Sandboxing](https://asciinema.org/a/625170)
- [AsciiCast: Network Sandboxing](https://asciinema.org/a/623664)
- [AsciiCast: Sandboxing Emacs with Syd](https://asciinema.org/a/627055)
- [Sydbox Manual Pages](https://man.exherbolinux.org/)
- [Sydbox profile for Nginx](https://gitlab.exherbo.org/sydbox/sydbox/-/raw/main/data/nginx.syd-3)
- [Sample remote shell, restricted at runtime with Sydbox](https://gitlab.exherbo.org/sydbox/sydbox/-/raw/main/lib/examples/go/goshell.go)
